### ISSUE TITLE
Record exists but cannot update values.

### ISSUE THREAD

**Issue by purplecandy** [tekartik/sembast.dart#34]
Trying to update records with custom key. I'm trying to update the "season" key inside my record.
It finds the records and returns a CastMap in the response.
But there is no change in the values.

The comments in the .update() says

>   /// If it does not exist, return null. if value is a map, keys with dot values
>   /// refer to a path in the map, unless the key is specifically escaped

**First I didn't get what does it means by dot values? How do I structure the path?**

**Secondly, what does it mean by key. Is it record key or store json key? If it means record key, then I'll probably need to update the entire json value right? Anyway to avoid that?**
```
var store = intMapStoreFactory.store(movieData.contentType);
      Map<String, dynamic> jsonValues = {"seasons": movieData.seasonObject};
      sembastDB.db.transaction((txn) async {
        var response =
            await store.record(int.parse(movieData.id)).update(txn, jsonValues);
        print('Response -> $response');
        print('Response -> ${response.runtimeType}');
      });
```
_This is my original record saved in the database_
```
{
  "id": "48866",
  "title": "The 100",
  "overview": "100 years in the future, when the Earth has been abandoned due to radioactivity, the last surviving humans live on an ark orbiting the planet — but the ark won't last forever. So the repressive regime picks 100 expendable juvenile delinquents to send down to Earth to see if the planet is still habitable.",
  "posterurl": "https://image.tmdb.org/t/p/w300//wBzNjurA8ijJPF21Ggs9nbviIzi.jpg",
  "posterpath": "/posters/48866.jpg",
  "backdropurl": "https://image.tmdb.org/t/p/w780//qYTIuJJ7fIehicAt3bl0vW70Sq6.jpg",
  "backdroppath": "/backdrops/48866.jpg",
  "latest_episode_air_date": "2019-05-07",
  "latest_episode_to_air": {
    "episode_number": 2,
    "id": 1686688
  },
  "season_length": 6,
  "episodes_length": "",
  "status": "Returning Series",
  "type": "",
  "episode_runtime": 43,
  "last_watched": "",
  "seasons": {
    "1": [
      1,
      2,
      3,
      4,
      5,
      6,
      7,
      8,
      9,
      10,
      11,
      12,
      13
    ],
  }
}
```

**Comment by alextekartik** [comment]
Sorry for the bad documentation. I made some updates here <https://github.com/tekartik/sembast.dart/blob/master/sembast/doc/writes.md#updating-fields>

The key is whatever you decide it to be. Either a fixed value you set or an auto-generated key you get when doing `store.add`

Can you try to reproduce your issue in a unit test? 
If I try the following (close to what you are doing), it works fine:

```dart
var store = intMapStoreFactory.store('movie');
Map<String, dynamic> jsonValues = {
  "seasons": {
    "1": [1, 2, 3]
  }
};
await db.transaction((txn) async {
  var map = {
    "id": "48866",
    "title": "The 100",
    "seasons": {
      "1": [1, 2, 3, 4, 5, 6],
    }
  };
  var key = await store.add(txn, map);
  var response = await store.record(key).update(txn, jsonValues);

  print('Response -> $response');
  print('Response -> ${response.runtimeType}');
  // only the seasons field has been updated
  expect(response, {
    'id': '48866',
    'title': 'The 100',
    'seasons': {
      '1': [1, 2, 3]
    }
  });
});
```

**Comment by purplecandy** [comment]
Hey I have found few issues. 
_There was some problem with my project/dependencies I could use dart test package but still I have managed to identify the problem_

```
I/flutter ( 6447): BASE RECORD -> {id: 12345, seasons: {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}}
I/flutter ( 6447): ADDING REDCORD

I/flutter ( 6447): Case 1: seasonObject.length > stored seasonObject.length
I/flutter ( 6447): seasonObject length -> 2
I/flutter ( 6447): Response length -> 2
I/flutter ( 6447): seasonObject -> {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
I/flutter ( 6447): Response -> {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}

I/flutter ( 6447): Case 2: seasonObject.length < stored seasonObject.length
I/flutter ( 6447): EXISTING ELEMENTS ARE BEING POPPED TO REDUCE LENGTH
I/flutter ( 6447): seasonObject length -> 1
I/flutter ( 6447): Response length -> 2
I/flutter ( 6447): seasonObject -> {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
I/flutter ( 6447): Response -> {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}

I/flutter ( 6447): Case 3: seasonObject.length > stored seasonObject.length
I/flutter ( 6447): ADDING WITH NEW KEY/VALUES
I/flutter ( 6447): seasonObject length -> 3
I/flutter ( 6447): Response length -> 3
I/flutter ( 6447): seasonObject -> {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
I/flutter ( 6447): Response -> {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 3: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}

I/flutter ( 6447): Case 4: seasonObject.length < stored seasonObject.length
I/flutter ( 6447): EXISTING ELEMENTS AT EACH KEY i.e(LIST<INT>) NUMBER OF VALUES ARE REDUCED TO ZERO i.e(LIST<INT>.length--)
I/flutter ( 6447): seasonObject length -> 3
I/flutter ( 6447): Response length -> 3
I/flutter ( 6447): seasonObject -> {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2: [1, 2, 3, 4, 5], 3: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
I/flutter ( 6447): Response -> {1: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 2: [1, 2, 3, 4, 5], 3: [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]}
```

**Conclusion**
Update works in - Case 1, Case 3, Case 4
Update doesn't work in- Case 2

In this issue I was using Case 2

_If the new record has a new key compared to the stored record it updates the value, else it does nothing_

_If the there is a modification within the values of records it updates the record._

I hope I explained it fine.

**Comment by alextekartik** [comment]
I think I understood the issue. Nice catch. Indeed there was an issue when merging inner map where the content was merged instead of replaced. I think I was able to reproduce it and fix it. Can you try again with the just published version 1.15.4?
Thanks

**Comment by purplecandy** [comment]
It's working properly now <3.
